rlclientlib/serialization/fb_serializer.h

```diff
@@ -0,0 +1,146 @@
+        evt.get_pass_prob());
+      return error_code::success;
+    }
+  };
+
+  template<>
+  struct fb_event_serializer<outcome_event> {
+    using fb_event_t = OutcomeEventHolder;
+    using offset_vector_t = std::vector<flatbuffers::Offset<fb_event_t>>;
+    using batch_builder_t = OutcomeEventBatchBuilder;
+
+    static size_t size_estimate(const outcome_event& evt) {
+      return  evt.get_event_id().size() +
+        evt.get_outcome().size() +
+        sizeof(evt.get_numeric_outcome());
+    }
+
+    static int serialize(outcome_event& evt, flatbuffers::FlatBufferBuilder& builder, flatbuffers::Offset<fb_event_t>& retval, api_status* status)
+    {
+      const auto event_id = builder.CreateString(evt.get_event_id());
+      switch(evt.get_outcome_type())
+      {
+        case outcome_event::outcome_type_string:
+        {
+          const auto outcome_str = builder.CreateString(evt.get_outcome());
+          const auto str_event = CreateStringEvent(builder, outcome_str).Union();
+          retval = CreateOutcomeEventHolder(builder, event_id, evt.get_pass_prob(), OutcomeEvent_StringEvent, str_event);
+          break;
+        }
+        case outcome_event::outcome_type_numeric:
+        {
+          const auto number_event = CreateNumericEvent(builder, evt.get_numeric_outcome()).Union();
+          retval = CreateOutcomeEventHolder(builder, event_id, evt.get_pass_prob(), OutcomeEvent_NumericEvent, number_event);
+          break;
+        }
+        case outcome_event::outcome_type_action_taken:
+        {
+          const auto action_taken_event = CreateActionTakenEvent(builder, evt.get_deferred_action()).Union();
+          retval = CreateOutcomeEventHolder(builder, event_id, evt.get_pass_prob(), OutcomeEvent_ActionTakenEvent, action_taken_event);
+          break;
+        }
+        default:
+        {
+          return report_error(status, error_code::serialize_unknown_outcome_type, error_code::serialize_unknown_outcome_type_s);
+        }
+      }
+
+      return error_code::success;
+    }
+  };
+
+  template <typename event_t>
+  struct fb_collection_serializer {
+    using serializer_t = fb_event_serializer<event_t>;
+    using buffer_t = utility::data_buffer;
+
+    static int message_id() { return message_type::UNKNOWN; }
+
+    fb_collection_serializer(buffer_t& buffer) :
+      _allocator(buffer),
+      _builder(buffer.body_capacity(), &_allocator),
+      _buffer(buffer)
+    {}
+
+    int add(event_t& evt, api_status* status = nullptr) {
+      flatbuffers::Offset<typename serializer_t::fb_event_t> offset;
+      RETURN_IF_FAIL(serializer_t::serialize(evt, _builder, offset, status));
+      _event_offsets.push_back(offset);
+      return error_code::success;
+    }
+
+    uint64_t size() const {
+      return _builder.GetSize();
+    }
+
+    void finalize() {
+      auto event_offsets = _builder.CreateVector(_event_offsets);
+      typename serializer_t::batch_builder_t batch_builder(_builder);
```

Comment:
Can we get rid of the batch builder. Flatbuffers compaction does not work well with our Event Hub batch sizing. One of my main reasons for supporting the preamble was to remove the necessity of flatbuffer batching. Can we have a preamble for each event?

Reply:
We added batching for throughput reasons.  Benchmarking EventHub REST API suggests that optimal payload size is around a 100k.  Performance drops off below and above that size.  
Can you elaborate on "Flatbuffers compaction does not work well with our Event Hub batch sizing"?

Reply:
I completely agree with batching the Event Hub message. The question comes down to batching the events at the level of the FlatBuffers or at the Event Hub message level. My concern was the fact that we couldn't truly know the fb_collection_serializer.size() until after fb_collection_serializer.finalize() is called. This makes the fill_buffer conditional less than perfect. My thought process was to remove the array (RankingEventBatch) , which was necessary due to lack of a preamble, and use multiple preambles to create our own array. Also, being inline with the future log format I see "[preamble][event][preamble][event]....." much simpler than "[preamble][X-events][preamble][Y-events]...". It not lost on me that after your refactor, everything is much cleaner and the  fb_collection_serializer.finalize() really shouldn't change the size much; if at all.

---
In reply to: [237304366](https://github.com/VowpalWabbit/reinforcement_learning/pull/24#discussion_r237304366) [](ancestors = 237304366)

Reply:
Will have to revisit this in the future.  Probably a considerable redesign of buffering pipeline to avoid memory reallocation and copies.  However having messgetype in the preamble will make this type of change low impact in the future.  The serverside already setup to handle either individual or batch message.